tensorflow/java/src/main/java/org/tensorflow/op/NameScope.java

```diff
@@ -0,0 +1,158 @@
+  }
+
+  private NameScope(String baseName, String baseOpName, Map<String, Integer> ids) {
+    this.baseName = baseName;
+    this.baseOpName = baseOpName;
+    if (ids != null) {
+      this.ids = ids;
+    } else {
+      this.ids = new HashMap<String, Integer>();
+    }
+  }
+
+  // Generate a unique name, different from existing ids.
+  //
+  // ids is a map from id to integer, representing a counter of the
+  // number of previous requests to generate a unique name for the
+  // given id.
+  //
+  // For instance, the first use of makeUnique("a") adds "a" -> 1
+  // to ids and returns "a".
+  //
+  // The second use of makeUnique("a") updates ids to "a" -> 2
+  // and returns "a_1", and so on.
+  private String makeUnique(String id) {
+    synchronized (ids) {
+      if (!ids.containsKey(id)) {
+        ids.put(id, 1);
+        return id;
+      } else {
+        int cur = ids.get(id);
+        ids.put(id, cur + 1);
+        return String.format("%s_%d", id, cur);
+      }
+    }
+  }
+
+  private String fullyQualify(String name) {
+    if (baseName != null) {
+      return String.format("%s/%s", baseName, name);
+    } else {
+      return name;
+    }
+  }
+
+  // If baseName is non-null, it is a prefix applied to all names
+  // created by this instance.
+  private final String baseName;
+
+  // If baseOpName is non-null, it is used to derive the unique name
+  // for operators rather than the provided default name.
+  private final String baseOpName;
+
+  // NameScope generates unique names by appending a numeric suffix if
+  // needed. This is a map containing names already created by this
+  // instance mapped to the next available numeric suffix for it.
+  private final Map<String, Integer> ids;
+
+  private static void checkPattern(Pattern pattern, String name) {
+    if (name == null) {
+      throw new IllegalArgumentException("Names cannot be null");
+    }
+    if (!pattern.matcher(name).matches()) {
+      throw new IllegalArgumentException(String.format("Invalid name '%s'", name));
+    }
+  }
+
+  // The constraints for operator and scope names originate from restrictions on node names
+  // noted in the proto definition core/framework/node_def.proto for NodeDef and actually
+  // implemented in core/framework/node_def_util.cc [Note that the proto comment does not include
+  // dash (-) in names, while the actual implementation permits it. These regexs follow the actual
+  // implementation.]
+  //
+  // These two patterns are used to ensure fully qualified names always start with a
+  // LETTER_DIGIT_DOT, followed by zero or more LETTER_DIGIT_DASH_DOT_SLASH_UNDERSCORE. SLASH is not
+  // permitted in actual user-supplied names to NameScope - it is used as a reserved character to
+  // separate subcomponents within fully qualified names.
+  private static final Pattern OP_NAME_REGEX = Pattern.compile("[A-Za-z0-9.][A-Za-z0-9_.\\-]*");
+  private static final Pattern SUBSCOPE_NAME_REGEX = Pattern.compile("[A-Za-z0-9_.\\-]+");
```

Comment:
This makes `"-"` s valid subscope name (`withSubScope("-")`). Is that intentional? (Wondering if we can use a single regexp for now.